src/lib.rs

```diff
@@ -1313,12 +1329,101 @@ mod test {
-        assert_eq!(format!("{}", Ratio::from_integer(-2)), "-2".to_string());
+    mod show {
+        use Ratio;
+
+        pub const _2: Ratio<isize> = super::_2;
+        pub const _1_2: Ratio<i32> = Ratio {
+            numer: 1,
+            denom: 2
+        };
+        pub const _0: Ratio<i64> = Ratio {
+            numer: 0,
+            denom: 1
+        };
+        pub const _NEG12: Ratio<isize> = Ratio {
+            numer: -12,
+            denom: 1
+        };
+        pub const _10_21: Ratio<i32> = Ratio {
+            numer: 10, 
+            denom: 21
+        };
+        pub const _3_47: Ratio<i64> = Ratio {
+            numer: 3,
+            denom: 47
+        };
+
+        #[test]
+        fn test_display() {
+            assert_eq!(format!("{}", _2), "2".to_string());
+            assert_eq!(format!("{}", _1_2), "1/2".to_string());
+            assert_eq!(format!("{}", _0), "0".to_string());
+            assert_eq!(format!("{}", _NEG12), "-12".to_string());
+        }
+
+        #[test]
+        fn test_octal() {
+            assert_eq!(format!("{:o}", _2), "2".to_string());
+            assert_eq!(format!("{:o}", _1_2), "1/2".to_string());
+            assert_eq!(format!("{:o}", _0), "0".to_string());
+            //assert_eq!(format!("{:o}", _NEG12), "-14".to_string());
+            assert_eq!(format!("{:o}", _10_21), "12/25".to_string());
+            assert_eq!(format!("{:#o}", _3_47), "0o3/0o57");
+        }
+
+        #[test]
+        fn test_lower_hex() {
+            assert_eq!(format!("{:x}", _2), "2".to_string());
+            assert_eq!(format!("{:x}", _1_2), "1/2".to_string());
+            assert_eq!(format!("{:x}", _0), "0".to_string());
+            //assert_eq!(format!("{:x}", _NEG12), "-c".to_string());
+            assert_eq!(format!("{:x}", _10_21), "a/15".to_string());
+            assert_eq!(format!("{:#x}", _3_47), "0x3/0x2f".to_string());
+        }
+
+        #[test]
+        fn test_upper_hex() {
+            assert_eq!(format!("{:X}", _2), "2".to_string());
+            assert_eq!(format!("{:X}", _1_2), "1/2".to_string());
+            assert_eq!(format!("{:X}", _0), "0".to_string());
+            //assert_eq!(format!("{:X}", _NEG12), "-C".to_string());
+            assert_eq!(format!("{:X}", _10_21), "A/15".to_string());
+            assert_eq!(format!("{:#X}", _3_47), "0x3/0x2F".to_string());
+        }
+
+        #[test]
+        fn test_binary() {
+            assert_eq!(format!("{:b}", _2), "10".to_string());
+            assert_eq!(format!("{:b}", _1_2), "1/10".to_string());
+            assert_eq!(format!("{:b}", _0), "0".to_string());
+            //assert_eq!(format!("{:b}", _NEG12), "-1100".to_string());
+            assert_eq!(format!("{:b}", _10_21), "1010/10101".to_string());
+            assert_eq!(format!("{:#b}", _3_47), "0b11/0b101111".to_string());
+        }
+        
+        // Integer types don't implement fmt::LowerExp / fmt::UpperExp.
```

Comment:
You could manually create some `Ratio<f64>` just for test purposes.
(at least until we get a proper type that works, perhaps `BigInt`.)

Reply:
The formatting actually requires an Integer and a One implementation, so f64 won't work right now anyway.

Reply:
AFAICS they're just `Eq + One + $imp`, no?

Reply:
By that reasoning, we could also try to cancel out the UpperExp and LowerExp scientific notation exponents. I thought about that, but I decided to keep this PR small.

Reply:
Oh right, but floats aren't `Eq`...

Reply:
If by "cancel out" you mean to actually divide them, that sounds pretty good.  Depending on #4 to get `f64`, then just write that in Exp.

If you agree that this would be desirable, then let's wait altogether on the Exp formats.

Reply:
Ok. I'll remove the Exp formats, we'll do them in another PR. And I'll make an alt form for floats that just doesn't check the denominator.

Reply:
I guess it could be more sophisticated than `f64` too, to support even larger exponents with `BigRational`.